torch/csrc/jit/passes/device_type_analysis.cpp

```diff
@@ -0,0 +1,211 @@
+#include <ATen/core/interned_strings.h>
+#include <ATen/core/jit_type.h>
+#include <c10/core/Device.h>
+#include <c10/util/ArrayRef.h>
+#include <c10/util/Optional.h>
+#include <torch/csrc/jit/ir/ir.h>
+#include <torch/csrc/jit/jit_log.h>
+#include <torch/csrc/jit/passes/device_type_analysis.h>
+#include <torch/library.h>
+#include <memory>
+#include <utility>
+
+namespace torch {
+namespace jit {
+
+namespace {
+
+using Tensor = at::Tensor;
+using Device = at::Device;
+
+using PropRule = std::function<bool(Node*)>;
+/*
+A Propagation Rule takes the Node, and
+applies the relevant properties to the Tensor outputs
+of the Node (based on the rule itself)
+
+Returns: Bool indicating if anything was changed
+*/
+
+bool setDeviceType(Value* value, c10::optional<Device> device) {
+  auto tensor_type = value->type()->cast<TensorType>();
+  TORCH_INTERNAL_ASSERT(tensor_type, "Expecting a tensor type");
+  bool changed = tensor_type->device() != device;
+  if (changed) {
+    value->setType(tensor_type->withDevice(device));
+  }
+  return changed;
+}
+
+bool setReturnsToDevice(Node* n, c10::optional<Device> device) {
+  bool changed = false;
+  for (Value* out : n->outputs()) {
+    auto tensor_type = out->type()->cast<TensorType>();
+    if (!tensor_type) {
+      continue;
+    }
+    changed |= setDeviceType(out, device);
+  }
+  return changed;
+}
+
+bool propWithNoDevice(Node* n) {
+  // Figure out what the common device to propagate is
+  // Types of tensors must match, except CPU zerodim, which any
+  // other type can overwrite
+
+  c10::optional<Device> device;
+  bool seen_any_device = false;
```

Comment:
do we need to have a `seen_any_device` boolean ? doesn't `c10::optional<Device> device == c10::nullopt` encode that?

Reply:
No, it does not. A `c10::nullopt` Device after device propagation means that a tensor can have multiple possible device_types. 

In this function, if you have already seen an input with a `c10::nullopt` device, you cannot safely assume a different tensor's device can overwrite that multiple possible device_types. On the other hand, if you haven't seen a device yet, you can overwrite that implicit `c10::nullopt` value.

Reply:
> . A c10::nullopt Device after device propagation means that a tensor can have multiple possible device_types.
Why not just bail as soon as we cannot infer? What is the point of retaining this information as local state?

Reply:
You asked me to overwrite previously infered types if we infer a different type/ can't infer a type. Setting the output as `c10::nullopt` is the way to "bail". 

I don't see the advantage of erroring out over marking it as `c10::nullopt`

Reply:
> You asked me to overwrite previously infered types if we infer a different type/ can't infer a type. Setting the output as c10::nullopt is the way to "bail".

Yea, that's not incompatible with returning from a function when we cannot infer an output instead of adding extra state that we have to keep track of. You could factor this part out into a `c10::optional<Device>` that is returned and then set on the output. 

As a general rule if you can minimize the amount of state / control flow a reader has to keep track of in function you should

Reply:
Having three different pieces of local state makes this pretty hard to review and reason about - you can bail on analysis while still overwriting the output. You just factor this part of the inference out to a function that returns `c10::optional<Device>` and then set the output equal to that value. Then you can delete `seen_any_device`.

Generally, if you can minimize the amount of state and control flow in a piece of code you should do so. Also it's generally performant to exit out of a loop as soon as you know the outcome instead of continuing to iterate.

Reply:
I think the root of the issue is that you think this function should exit as soon as there is only one valid output device that we can infer. I think that is ok if we are manually whitelisting ops for Device Type Analysis (which you seem to be very explicitly against). 

The default inference I have written instead only infers if all inputs have the same device (except CPU zerodim) in order to be more conservative and checks for the case of mismatching device types, as I don't think we can by assume that we have made custom cases for all of the ops that can mix device types.